File: app/src/main/java/breakingumbrella/connectit/entity/gameobjects/HeatMap.java
package breakingumbrella.connectit.entity.gameobjects;

public class HeatMap {

	private int[][] heatCells;
	private Figure figure = new Figure();

	public HeatMap(GameField gameField) {
		updateHeatMap(gameField);
	}

	public HeatMap(int xSize, int ySize) {
		for(int i = 0; i < xSize; i++) {
			for(int j = 0; j < ySize; j++) {
				heatCells = new int[xSize][ySize];
			}
		}
	}

	public boolean isHeatMapIsEmpty() {
		boolean isEmpty = true;
		for(int i = 0; i < heatCells.length; i++) {
			for (int j = 0; j < heatCells[0].length; j++) {
				if(heatCells[i][j] > 0) {
					isEmpty = false;
				}
			}
		}
		return false;
	}

	public int[] getFirstHotCell() {
		int[] positon = {0, 0};
		for(int i = 0; i < heatCells.length; i++) {
			for(int j = 0; j < heatCells[0].length; j++) {
				if(heatCells[i][j] > 0) {
					positon[0] = i;
					positon[1] = j;
				}
			}
		}
		return positon;
	}

	public boolean isCellIsHot(int i, int j) {
		return heatCells[i][j] == 0 ;
	}

	public void add(int x, int y, int value) {
		heatCells[x][y] = value;
	}

	public void updateHeatMap(GameField gameField) {
		heatCells = new int[gameField.getSizeX()][gameField.getSizeY()];
		int cellHeat = 0;
		for(int i = 0; i < gameField.getSizeX(); i++) {
			for(int j = 0; j < gameField.getSizeY(); j++) {
				if(gameField.getFigureAtPosition(i, j).isEmpty()) {
					cellHeat += updateRight(gameField, i, j);
					cellHeat += updateLeft(gameField, i, j);
					cellHeat += updateTop(gameField, i, j);
					cellHeat += updateBottom(gameField, i, j);
					cellHeat += updateLowerLeftCorner(gameField, i, j);
					cellHeat += updateLowerRightCorner(gameField, i, j);
					cellHeat += updateUpperLeftCorner(gameField, i, j);
					cellHeat += updateUpperRightCorner(gameField, i, j);
					if (cellHeat > 0) {
						heatCells[i][j] = cellHeat;
					}
					cellHeat = 0;
				}
			}
		}
	}

	private int updateRight(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i, j + 1);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}

	private int updateLeft(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i, j - 1);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}

	private int updateTop(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i - 1, j);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}

	private int updateBottom(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i + 1, j);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}

	private int updateUpperRightCorner(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i - 1, j + 1);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}

	private int updateUpperLeftCorner(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i - 1, j - 1);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}

	private int updateLowerLeftCorner(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i + 1, j - 1);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}

	private int updateLowerRightCorner(GameField gameField, int i, int j) {
		figure = gameField.getFigureAtPosition(i + 1, j + 1);
		if(figure.isEmpty() == false && figure.isFigureCrossed() == false) {
			return 1;
		}
		return 0;
	}
}

File: app/src/main/java/breakingumbrella/connectit/entity/gamelogic/ai/HeatMap.kt
package breakingumbrella.connectit.entity.gamelogic.ai

import breakingumbrella.connectit.entity.gameobjects.Figure
import breakingumbrella.connectit.entity.gameobjects.GameField

class HeatMap(gameField: GameField) : Iterator<IntArray> {

    init {
        updateHeatMap(gameField)
    }

    private data class Cell(val heat: Int, val position: IntArray)

    private lateinit var heatCells: MutableList<Cell>

    private lateinit var iterator: ListIterator<Cell>

    private var figure = Figure()

    fun isHeatMapIsEmpty(): Boolean {
        return heatCells.isEmpty()
    }

    fun getFirstHotCell(): IntArray? {
        sort()
        return heatCells.first().position
    }

    fun sort() {
        heatCells = heatCells.sortedBy { it.heat } .toMutableList()
    }

    fun updateHeatMap(gameField: GameField) {
        heatCells = mutableListOf()
        iterator = heatCells.listIterator()
        var cellHeat = 0
        for (i in 0 until gameField.sizeX) {
            for (j in 0 until gameField.sizeY) {
                if (gameField.getFigureAtPosition(i, j).isEmpty) {
                    cellHeat += updateRight(gameField, i, j)
                    cellHeat += updateLeft(gameField, i, j)
                    cellHeat += updateTop(gameField, i, j)
                    cellHeat += updateBottom(gameField, i, j)
                    cellHeat += updateLowerLeftCorner(gameField, i, j)
                    cellHeat += updateLowerRightCorner(gameField, i, j)
                    cellHeat += updateUpperLeftCorner(gameField, i, j)
                    cellHeat += updateUpperRightCorner(gameField, i, j)
                    if (cellHeat > 0) {
                        heatCells.add(Cell(cellHeat, intArrayOf(i, j)))
                    }
                    cellHeat = 0
                }
            }
        }
    }

    //===================ITERATOR==================//

    private var currentIndex = 0

    override fun hasNext(): Boolean {
        return currentIndex < heatCells.size
    }

    override fun next(): IntArray {
        return heatCells[currentIndex++].position
    }

    //===================ITERATOR==================//

    private fun updateRight(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i, j + 1)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

    private fun updateLeft(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i, j - 1)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

    private fun updateTop(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i - 1, j)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

    private fun updateBottom(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i + 1, j)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

    private fun updateUpperRightCorner(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i - 1, j + 1)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

    private fun updateUpperLeftCorner(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i - 1, j - 1)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

    private fun updateLowerLeftCorner(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i + 1, j - 1)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

    private fun updateLowerRightCorner(gameField: GameField, i: Int, j: Int): Int {
        figure = gameField.getFigureAtPosition(i + 1, j + 1)
        return if (figure.isEmpty == false && figure.isFigureCrossed == false) {
            1
        } else 0
    }

}